### ISSUE TITLE
Double-free if easy is freed at global destruction

### ISSUE THREAD

**Issue by FGasper** [sparky/perl-Net-Curl#50]
```
> perl -Ilib -MNet::Curl::Promiser::AnyEvent -MAnyEvent -Mstrict -w -MNet::Curl::Easy -e'our $easy = Net::Curl::Easy->new(); $easy->setopt( Net::Curl::Easy::CURLOPT_URL, "https://google.com" ); print qq<$INC{"Net/Curl/Easy.pm"}$/>; my $promiser = Net::Curl::Promiser::AnyEvent->new(); $promiser->add_handle($easy);'
/opt/local/lib/perl5/site_perl/5.26/darwin-thread-multi-2level/Net/Curl/Easy.pm
perl(24346,0x1169935c0) malloc: *** error for object 0x7fc16fd3abe0: pointer being freed was not allocated
perl(24346,0x1169935c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```
Also happens with Mojo:
```
> perl -Ilib -MNet::Curl::Promiser::Mojo -MAnyEvent -Mstrict -w -MNet::Curl::Easy -e'our $easy = Net::Curl::Easy->new(); $easy->setopt( Net::Curl::Easy::CURLOPT_URL, "https://google.com" ); print qq<$INC{"Net/Curl/Easy.pm"}$/>; my $promiser = Net::Curl::Promiser::Mojo->new(); $promiser->add_handle($easy);'
/opt/local/lib/perl5/site_perl/5.26/darwin-thread-multi-2level/Net/Curl/Easy.pm
perl(24382,0x10d8095c0) malloc: *** error for object 0x7f84cff55bb0: pointer being freed was not allocated
perl(24382,0x10d8095c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```
Seen on Perl 5.26 on macOS 10.14.

I think I have a fix; I’m just writing a test.

**Comment by FGasper** [comment]
Simplest repro:
```
> perl -Mstrict -w -MNet::Curl::Multi -MNet::Curl::Easy -e'my $multi = Net::Curl::Multi->new(); our $easy = Net::Curl::Easy->new(); $multi->add_handle($easy);'
perl(25069,0x10d0575c0) malloc: *** error for object 0x7fa152c0bc90: pointer being freed was not allocated
perl(25069,0x10d0575c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

**Comment by creaktive** [comment]
Thanks! Fix being tested :) http://matrix.cpantesters.org/?dist=Net-Curl+0.46_1

**Comment by FGasper** [comment]
@creaktive 4 days without a failure … sufficient for a release?

**Comment by creaktive** [comment]
Released [Net::Curl 0.46](https://metacpan.org/release/SYP/Net-Curl-0.46)!
